ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifier.java

```diff
@@ -90,8 +93,31 @@ public void onAttestationsDue(final UInt64 slot) {
     eventThread.execute(() -> internalAttestationsDue(slot));
   }
 
-  public SafeFuture<Optional<Bytes8>> getPayloadId() {
-    return eventThread.execute(() -> lastPayloadId);
+  public SafeFuture<Optional<Bytes8>> getPayloadId(Bytes32 beaconBlockRoot, UInt64 slot) {
+    return eventThread.execute(
+        () -> {
+          if (lastPayloadId.isPresent()) {
+            if (lastPayloadIdBeaconBlockRoot
+                    .map(fcsBeaconRoot -> fcsBeaconRoot.equals(beaconBlockRoot))
+                    .orElse(false)
+                && lastPayloadIdSlot.map(fcsSlot -> fcsSlot.compareTo(slot) == 0).orElse(false)) {
```

Comment:
I'd suggest comparing the execution hashes instead.  You can get the last payload Id head execution hash from `lastSentForkChoiceState` and convert the beaconBlockRoot to it's execution hash via `recentChainData.getForkChoiceStrategy().orElseThrow().executionBlockHash(beaconBlockRoot)` (though it's probably worth adding a `RecentChainData.getExecutionHashForBlockRoot` method to wrap that more nicely in the same way we have `RecentChainData.getSlotForBlockRoot`.